src/Toml/Parser/Value.hs

```diff
@@ -41,15 +45,33 @@ decimalP = zero <|> more
     check :: Maybe Integer -> Parser Integer
     check = maybe (fail "Not an integer") pure
 
+-- | Parser for hexadecimal, octal and binary numbers : included parsing
+numberP :: Parser Integer -> Parser Char -> Parser Integer
+numberP parser parseDigit = more
+  where
+    more :: Parser Integer
+    more = check =<< intValueMaybe . concat <$> sepBy1 (some parseDigit) (char '_')
+
+    intValueMaybe :: String -> Maybe Integer
+    intValueMaybe = parseMaybe (parser :: Parser Integer) . fromString
```

Comment:
Let's rename `parser` to `parseInteger`. Also, the type annotation here is not needed, since the type of `parse` is already specified in the top-level signature.